hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/container/ozoneimpl/TestOzoneContainerWithTLS.java

```diff
@@ -18,357 +18,374 @@
+  @Test
+  public void testLongLivingClientWithCertRenews() throws Exception {
+    LogCapturer logs = captureLogs(getLogger(ClientTrustManager.class));
+    OzoneContainer container = createAndStartOzoneContainerInstance();
 
-      List<DatanodeDetails> sourceDatanodes = new ArrayList<>();
-      sourceDatanodes.add(dn);
-      if (containerTokenEnabled) {
-        // old client still function well after certificate expired
-        Token<ContainerTokenIdentifier> token = secretManager.generateToken(
-            UserGroupInformation.getCurrentUser().getUserName(),
-            ContainerID.valueOf(containerId));
-        createSecureContainer(client, containerId, token);
-        closeSecureContainer(client, containerId++, token);
-      } else {
-        createContainer(client, containerId);
-        closeContainer(client, containerId++);
-      }
+    ScmClientConfig scmClientConf = conf.getObject(ScmClientConfig.class);
+    scmClientConf.setStaleThreshold(500);
+    XceiverClientManager clientManager =
+        new XceiverClientManager(conf, scmClientConf, aClientTrustManager());
+    assertClientTrustManagerLoading(true, logs, "Client loaded certificates.");
 
-      // Download newly created container will fail because of cert expired
-      GenericTestUtils.LogCapturer logCapture = GenericTestUtils.LogCapturer
-          .captureLogs(SimpleContainerDownloader.LOG);
-      SimpleContainerDownloader downloader =
-          new SimpleContainerDownloader(conf, caClient);
-      Path file = downloader.getContainerDataFromReplicas(
-          containerId, sourceDatanodes, null, NO_COMPRESSION);
-      downloader.close();
-      Assert.assertNull(file);
-      Assert.assertTrue(logCapture.getOutput().contains(
-          "java.security.cert.CertificateExpiredException"));
+    XceiverClientSpi client = null;
+    try {
+      client = clientManager.acquireClient(pipeline);
+      createAndCloseContainer(client, false);
+      assertClientTrustManagerLoading(false, logs,
+          "Check client did not reloaded certificates.");
+
+      letCACertExpire();
+
+      // works based off of initial SSL handshake
+      createAndCloseContainer(client, false);
+      assertClientTrustManagerLoading(false, logs,
+          "Check client did not reloaded certificates.");
+
+      clientManager.releaseClient(client, true);
+      client = clientManager.acquireClient(pipeline);
+      assertClientTrustManagerLoading(false, logs,
+          "Check second client creation does not reload certificates.");
+      // aim is to only load certs at first start, and in case of a verification
+      // failure
+
+      try {
+        // fails after a new client is created, as we just have expired certs
+        // in the source (caClient), so container ops still fail even though
+        // a retry/reload happens, but we expect it to happen before the failure
+        createAndCloseContainer(client, false);
+        assertClientTrustManagerFailedAndRetried(logs);
```

Comment:
This assertClientTrustManagerFailedAndRetried shall be moved to catch scope, since createAndCloseContainer will throw exception.

Reply:
Yes you are right, moved it into the catch clause.